### PR TITLE
Add shader code (re)loading

### PR DESCRIPTION
**Pull request by alteous**
This PR prepares for a more sophisticated shader system as discussed in #28.

The window creation API is greatly simplified for simple use cases.

### CODE REVIEW

File: src/window.rs [hunk 1]
```diff
@@ -113,16 +109,17 @@ impl Builder {
 }
 
 impl Window {
+    /// Create a new window with default parameters.
+    pub fn new<T: Into<String>>(title: T) -> Self {
```
Comment (vitvakatu) [inline src/window.rs]: 👍

File: src/render.rs [hunk 2]
```diff
@@ -296,6 +226,206 @@ struct DebugQuad {
...
+
+    /// Corresponds to `Material::MeshPbr`.
+    pbr: gfx::PipelineState<back::Resources, pbr_pipe::Meta>,
+
+    /// Used internally for rendering `Background::Skybox`.
+    skybox: gfx::PipelineState<back::Resources, quad_pipe::Meta>,
+}
+
+impl PipelineStates {
+    pub fn init(src: &ProgramSourceSet, factory: &mut back::Factory) -> Self {
```
Comment (alteous) [inline src/render.rs]: The idea here is if the `ProgramSourceSet` needs to change then we can reload *all* the shaders in one go.

File: src/render.rs [hunk 3]
```diff
@@ -331,35 +452,104 @@ pub struct Renderer {
...
 impl Renderer {
     #[cfg(feature = "opengl")]
-    #[doc(hidden)]
-    pub fn new(
+    pub(crate) fn new(
         builder: glutin::WindowBuilder,
         context: glutin::ContextBuilder,
         event_loop: &glutin::EventsLoop,
-        shader_path: &Path,
+        source: &ProgramSourceSet,
```
Comment (alteous) [inline src/render.rs]: Preparing shaders is now delegated to the caller.

File: build.rs [hunk 4]
```diff
@@ -0,0 +1,10 @@
+extern crate includedir_codegen;
+
+use includedir_codegen::Compression;
+
+fn main() {
+    includedir_codegen::start("FILES")
+        .dir("data/shaders", Compression::None)
```
[the remark this answers is not in the document]
Comment (alteous) [reply build.rs]: It's so little data I didn't think it would be worth it. Would gzip be adequate?

**Comment by alteous** [comment]
I forgot about the overriding scheme actually! Will fix in the next commit. Auto-reload will be delegated to the user.

File: src/render.rs [hunk 5]
```diff
@@ -331,35 +452,104 @@ pub struct Renderer {
...
+    /// Load the named shader from the default set of shaders.
+    pub fn default(name: &str, suffix: &str) -> io::Result<Self> {
+        let path = format!("data/shaders/{}_{}.glsl", name, suffix);
+        let unprocessed = data::FILES.get(&path).unwrap();
+        let processed = Self::preprocess("", str::from_utf8(unprocessed.borrow()).unwrap())?;
+        Ok(ShaderSource { code: processed })
+    }
+
+    /// Load the named shader from the given directory path.
+    pub fn user(root: &str, name: &str, suffix: &str) -> io::Result<Self> {
```
Comment (vitvakatu) [inline src/render.rs]: I would like to see `&Path`s here, not `&str`.

**Comment by alteous** [comment]
I implore you to try out the `reload` example! :smile:

**Comment by alteous** [comment]
The shader code was adapted from http://nuclear.mutantstargoat.com/articles/sdr_fract/ (public domain).

**Comment by alteous** [comment]
Should be ready to merge now.

File: examples/reload.rs [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+            let source_set = three::render::source::Set {
+                sprite: three::render::source::Sprite::user(&dir).unwrap(),
+                ..Default::default()
+            };
+            match three::render::PipelineStates::new(&source_set, &mut win.factory) {
+                Ok(pipeline_states) => {
+                    win.renderer.reload(pipeline_states);
+                    reload = false;
+                }
+                Err(err) => println!("{:#?}", err),
```
Comment (kvark) [inline examples/reload.rs]: leaving `reload==true` would mean the attempts to reload happen each successive frame, accompanied by the `println` spam. I suppose we need to reset `reload` flag regardless?

File: src/render/mod.rs.bk [hunk 7]
```diff
@@ -0,0 +1,957 @@
+use cgmath::{Matrix4, SquareMatrix, Transform as Transform_, Vector3};
```
Comment (kvark) [inline src/render/mod.rs.bk]: this file does not belong here

File: src/render/source.rs [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+        root: P,
+        code: &str,
+    ) -> io::Result<String> {
+        let root = root.as_ref();
+        let mut new_code = String::new();
+        for line in code.lines() {
+            if line.starts_with("#include") {
+                for dep_name in line.split(' ').skip(1) {
+                    match dep_name {
+                        "locals" | "lights" | "globals" => {
```
Comment (kvark) [inline src/render/source.rs]: hmm, I'd prefer us to have a solid distinction of the built-in versus custom includes in shaders
something like `locals` for our shaders and `/locals` for user shaders

File: src/window.rs [hunk 9]
```diff
@@ -90,16 +98,58 @@ impl Builder {
...
         let context = glutin::ContextBuilder::new()
             .with_vsync(self.vsync)
             .with_multisampling(self.multisampling);
 
+        let mut source_set = render::source::Set::default();
+        if let Some(path) = self.shader_directory.as_ref() {
+            let path = path.to_str().unwrap();
+            macro_rules! try_override {
+                ($name:ident) => {
+                    if let Ok(src) = render::Source::user(
```
Comment (kvark) [inline src/window.rs]: nit: let's use `match` against `if let ... else` here and below

File: examples/reload.rs (hunk 6 above)
Comment (alteous) [reply examples/reload.rs]: That's an oversight, thanks.

File: src/window.rs (hunk 9 above)
Comment (alteous) [reply src/window.rs]: Much better idea.

File: src/render/source.rs (hunk 8 above)
Comment (alteous) [reply src/render/source.rs]: Is that not covered already? :confused:

**Comment by alteous** [comment]
Hooray! :fireworks:

**Pull request merged**